### PR TITLE
Subscription model

### PR DESCRIPTION
**Pull request by Fastjur**
For https://github.com/AreaFiftyLAN/lancie-frontend/pull/521

Last thing todo for subscription model. Get a clean error message for unique constraint.
image

## Todo
- [x] Add tests
- [x] Test the remove/unsubscribe functionality (Perhaps this should be done using an id after all, as we'd like this to be able to be a clickable `unsubscribe` link as per CAN-SPAM act)

### CODE REVIEW

File: src/main/java/ch/wisv/areafiftylan/extras/mailupdates/controller/SubscriptionController.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.util.Collection;
+
+import javax.websocket.server.PathParam;
```
Comment (martijnjanssen) [inline src/main/java/ch/wisv/areafiftylan/extras/mailupdates/controller/SubscriptionController.java]: This should not be here, maybe an `Ctrl+Alt+o` to clean up the imports?

Comment (Fastjur) [reply src/main/java/ch/wisv/areafiftylan/extras/mailupdates/controller/SubscriptionController.java]: that's weird...

File: src/main/java/ch/wisv/areafiftylan/extras/mailupdates/service/SubscriptionServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+  @Override
+  public Collection<Subscription> getSubscriptions() {
+    return subscriptionRepository.findAll();
+  }
+
+  @Override
+  public Subscription getSubscriptionByEmail(String email) {
+    Optional<Subscription> subscription = subscriptionRepository.findByEmail(email);
+    if (!subscription.isPresent()) {
```
Comment (martijnjanssen) [inline src/main/java/ch/wisv/areafiftylan/extras/mailupdates/service/SubscriptionServiceImpl.java]: I think we can use an `orElseThrow` here, which does exactly what you are trying to do here. [Check it out.](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#orElseThrow-java.util.function.Supplier-)

Comment (Fastjur) [reply src/main/java/ch/wisv/areafiftylan/extras/mailupdates/service/SubscriptionServiceImpl.java]: I tried that but it wouldn't work. It was the first thing I wrote, but I got some funky error messages. I am too tired now to fix it though, I'll probably find the issue tomorrow.

File: src/main/java/ch/wisv/areafiftylan/extras/mailupdates/controller/SubscriptionController.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    this.subscriptionService = subscriptionService;
+  }
+
+  @GetMapping
+  @PreAuthorize("hasRole('COMMITTEE')")
+  public Collection<Subscription> getAllSubscriptions() {
+    return subscriptionService.getSubscriptions();
+  }
+
+  @GetMapping("/{id}")
```
Comment (martijnjanssen) [inline src/main/java/ch/wisv/areafiftylan/extras/mailupdates/controller/SubscriptionController.java]: Why would we need an individual email?

**Comment by Fastjur** [comment]
Ah I see you found the issue for throwing the error too.

I have removed some unnecessary code too, we only need to add and remove by email and retrieve all subscriptions

**Comment by Fastjur** [comment]
Woops. I fucked up the request param. But CH is closing now. Will fix when home :P

**Comment by Fastjur** [comment]
So. To get back on the error messages. It appears that in the user controller we also simply send `ex.getMessage()`

How are these nice and readable errors, while they are not here?

File: src/main/java/ch/wisv/areafiftylan/extras/mailupdates/model/Subscription.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+ * Used to gather emails from the frontend, to allow interested users
+ * to receive emails about the event.
+ *
+ * @author Jurriaan Den Toonder Created on 23-10-17
+ */
+@Entity
+@Data
+@RequiredArgsConstructor
+@NoArgsConstructor
+public class Subscription {
```
Comment (martijnjanssen) [inline src/main/java/ch/wisv/areafiftylan/extras/mailupdates/model/Subscription.java]: I just thought about this, it is very easy to allow users to unsubscribe, just add an active flag to the subscription.

Comment (Fastjur) [reply src/main/java/ch/wisv/areafiftylan/extras/mailupdates/model/Subscription.java]: Discussed, keeping it like this

**Comment by Fastjur** [comment]
I have tested the migrations. They are working! *Ready to merge*

**Pull request merged**